# Unsupervised Deep Topology Embedded Characterization of Single-Cell Chromatin Accessibility Profiles

`scDTEC` is a python package for classifying cell clusters based on chromatin accessibility from scATAC-seq using a deep clustering model based on graph neural networks.

- [Overview](#overview)
- [System Requirments](#system-requirments)
- [Installation Guide](#installation-guide)
- [Usage](#Usage)
- [Data Availability](#data-availability)
- [License](#license)

# Overview

Cell clustering plays a crucial role in the analysis of single-cell Assay for Transposase-Accessible Chromatin using sequencing (scATAC-seq) data, where chromatin features unveil intercellular variability in gene regulation. Single-cell deep embedding models have gained significant popularity for learning feature representations in a low-dimensional space to facilitate clustering. However, these models are susceptible to technical artifacts, noise, and missing values, which can adversely affect their overall performance. To address those limitations, we propose the \textbf{S}ingle-\textbf{C}ell \textbf{D}eep \textbf{T}opology \textbf{E}mbedded \textbf{C}haracterization (scDTEC) model, which obtains a low-dimensional fused representation of chromatin accessibility profiles and cell topological information. scDTEC employs a topology variational autoencoder to transform high-dimensional input data into latent layer representations that combine chromatin accessibility profiles with cellular topological information and then generate reconstructed chromatin information on the decoder. Concurrently, scDTEC employs a contrastive loss to maximize the consistency between the anchor graph derived from the row data and the learning graph generated by the graph learner model and uses the anchor graph as the learning objective. Finally, scDTEC employs a joint optimization paradigm to simultaneously optimize the embedding of cell fusion information and the updating of cell topology structure, guiding the precise partitioning of cell clusters. The results of our evaluation demonstrate the superiority of scDTEC over a variety of cutting-edge methods.
![scDTEC](https://github.com/jingtairan/scDTEC/blob/master/Framework.png)

## NOTICE

Due to the capacity limiation of Github, we put the relevant files (all datasets, model PT files and latent results) in our `<a href="https://figshare.com/articles/dataset/scDTEC/24591390">`FigShare`</a>`. All source code, data and model are open source and can be downloaded from GitHub or any other processes.

# System Requirments

## Hardware requirements

`scDTEC` package requires only a standard computer with enough RAM to support the in-memory operations.

## Software requirenebts

### OS Requirments

This package is supported for *Linux*. The package has been tested on the following systems:

+ Linux: Ubuntu 20.04

### Python Dependencies

`scDTEC` mainly depends on the Python scientific stack.

```
numpy
annadata
scipy
scanpy
scikit-learn
pytorch
dgl
munkres
```

For specific setting, please see `<a href="https://github.com/jingtairan/scDTEC/blob/master/requirements.txt">`requirements`</a>` or `<a href="https://github.com/jingtairan/scDTEC/blob/master/requirements.yml">`yml`</a>`.

# Installation Guide:

### We recommend using a conda environment to build scDTEC.

```
$ conda env create -f scDTEC.yml
```

# Usage

First all datasets to be trained should be downloaded and it is recommended to construct a `.sclaex` folder at the upper level of the code folder and put the datasets into it, after which the model can be trained with one of the scATAC-seq datasets using the following commands:

```python
python main.py -d Forebrain -k 8 --min_peaks 2 --min_cells 50
```

where k is the number of categories in the dataset, and min_peaks and min_cells are parameters set for data preprocessing, and the corresponding settings can be found in the paper.

We provide pretrain_model and pretrain_graphlearner parameters for users to call the saved model parameters directly.

```python
python main.py -d Forebrain -k 8 --min_peaks 2 --min_cells 50 --pretrain_model "/output/scDGEC_Forebrain_model.pt" --pretrain_graphlearner "/output/scDGEC_Forebrain_graphlearner.pt"
```

At the end of the program run four files are output, which are

```
scDTEC_dataset_model.pt
scDTEC_dataset_graphlearner.pt
scDTEC_dataset_adata.h5ad
scDTEC_dataset.png
```

The above are examples of training a model with the Forebrain dataset as an example or dividing the dataset directly using a pre-trained model.

# Data Availability

All supporting source code and data can be downloaded from `<a href="https://github.com/jingtairan/scDTEC">`here`</a>` and `<a href="https://figshare.com/articles/dataset/scDTEC/24591390">`FigShare`</a>`.

# License

This project is covered under the **MIT License**.

Thank you for using scDTEC! Any questions, suggestions or advice are welcome!
Contact:  yichang@jlu.edu.cn, lixt314@jlu.edu.cn, jingtr21@mails.jlu.edu.cn
